### PR TITLE
rpc generate: print useful help and error message

### PR DESCRIPTION
**Pull request by jonatack**
This was a requested follow-up to #19133 and #17700 to alleviate confusion and head-scratching by people following tutorials that use `generate`. See https://github.com/bitcoin/bitcoin/pull/19455#issuecomment-668172916 below, https://github.com/bitcoin/bitcoin/pull/19133#issuecomment-636860943 and https://github.com/bitcoin/bitcoin/pull/17700#issuecomment-566159096.

before
```
$ bitcoin-cli help generate
help: unknown command: generate

$ bitcoin-cli generate
error code: -32601
error message:
Method not found
```

after
```
$ bitcoin-cli help generate
generate ( nblocks maxtries ) has been replaced by the -generate cli option. Refer to -help for more information.

$ bitcoin-cli generate
error code: -32601
error message:
generate ( nblocks maxtries ) has been replaced by the -generate cli option. Refer to -help for more information.
```

In the general help it remains hidden, as requested by laanwj.
```
$ bitcoin-cli help

== Generating ==
generateblock "output" ["rawtx/txid",...]
generatetoaddress nblocks "address" ( maxtries )
generatetodescriptor num_blocks "descriptor" ( maxtries )
```

### CODE REVIEW

**Comment by jonatack** [comment]
Tagging @adamjonas and @MarcoFalke who suggested this feature.

**Comment by adamjonas** [comment]
Concept ACK. Thanks for circling back on this.

When looking at the help options (`bitcoin-cli -regtest help`), this explanation looks out of place. I'd suggest rewording to at least start with the word generate to keep it alphabetical and consistent.

Screen Shot 2020-07-06 at 7 51 59 PM

**Comment by jonatack** [comment]
> When looking at the help options (`bitcoin-cli -regtest help`), this explanation looks out of place. I'd suggest rewording to at least start with the word generate to keep it alphabetical and consistent.

Thanks Jonas! Good point. Updated to the following, let me know what you think.
```
== Generating ==
generate ( nblocks maxtries ) has been replaced by the -generate cli option. Refer to -help for more information.
generateblock "output" ["rawtx/txid",...]
generatetoaddress nblocks "address" ( maxtries )
generatetodescriptor num_blocks "descriptor" ( maxtries )
```

**Comment by jonatack** [comment]
Moved the test to its own file to hopefully appease the RPC coverage checker.

File: src/rpc/mining.cpp [hunk 1]
```diff
@@ -1194,6 +1205,7 @@ static const CRPCCommand commands[] =
     { "generating",         "generatetoaddress",      &generatetoaddress,      {"nblocks","address","maxtries"} },
     { "generating",         "generatetodescriptor",   &generatetodescriptor,   {"num_blocks","descriptor","maxtries"} },
     { "generating",         "generateblock",          &generateblock,          {"output","transactions"} },
+    { "generating",         "generate",               &generate,               {} },
```
Comment (laanwj) [inline src/rpc/mining.cpp:1208]: Why not make this a hidden RPC?

**Comment by laanwj** [comment]
I'm not sure on this. On one hand this is useful, on the other hand we did remove the similar message for `getinfo` a long time ago in b2f23c41538eaadd71c373ada75dd3a982eeb8bf.  `generate`, though shorter ago,  was removed more than a year ago (in 8bb3e4c487500a580e3e18791b1f4e7dcdd35442), so it seems to be a bit late to add this.
But no strong opposition.

File: src/rpc/mining.cpp (hunk 1 above)
Comment (jonatack) [reply src/rpc/mining.cpp]: Yes, that's better. Done.

**Comment by jonatack** [comment]
No ACKs. Closing.

**Comment by adamjonas** [comment]
This is my fault for the slow review, but I still think it's a useful addition to reduce confusion for all the outdated tutorials out there  [[1](https://bitcoinj.github.io/testing#:~:text=generate%20101), [2](https://bitcoin.stackexchange.com/questions/77389/bitcoin-regtest-mining#:~:text=generate), [3](https://www.oreilly.com/library/view/mastering-blockchain-/9781788839044/c0bf052c-b899-4a86-b8d4-86a055b1eae7.xhtml#:~:text=generate), [4](http://diyhpl.us/wiki/transcripts/austin-bitcoin-developers/2018-08-17-richard-bondi-bitcoin-cli-regtest/#:~:text=generate%201)].

utACK f0aa8aeea5a183ea44a877255d12db7732f2e0a8

Tried `help generate`, which returns a helpful prompt:
```
$ bitcoin-cli help generate
generate ( nblocks maxtries ) has been replaced by the -generate cli option. Refer to -help for more information.
```

This is much more instructive than the method not found error.
```
$ bitcoin-cli generate
error code: -32601
error message:
generate ( nblocks maxtries ) has been replaced by the -generate cli option. Refer to -help for more information.
```

And verified that it is a hidden RPC command
Screen Shot 2020-08-03 at 2 15 02 PM

I'm still an advocate for including this for the poor, unfortunate souls out there just trying to get started on regtest.

**Comment by jonatack** [comment]
Thanks for the review @adamjonas. Re-opening to see if this gets traction.

**Comment by kallewoof** [comment]
Concept / code review ACK. LGTM

No strong opinion, but an alternative might be to PR an update on those outdated docs, and/or ping the respective authors. Perhaps both; ping and merge.

**Comment by jonatack** [comment]
Thanks @adamjonas and @kallewoof :heart:

Good idea; I notified on the bitcoin.org github a couple months ago that `generate` is back as a cli option, though following up on all the tuturials would be helpful for people. 

If the maintainers agree on this change, ISTM the 2 ACKs and the test coverage here may enough to merge this no-risk user-facing aid.

File: test/functional/test_runner.py [hunk 2]
```diff
@@ -708,14 +709,16 @@ def _get_uncovered_rpc_commands(self):
...
         reference_filename = 'rpc_interface.txt'
         coverage_file_prefix = 'coverage.'
 
         coverage_ref_filename = os.path.join(self.dir, reference_filename)
         coverage_filenames = set()
         all_cmds = set()
-        covered_cmds = set()
+        # Consider RPC generate covered, because it is overloaded in
+        # test_framework/test_node.py and not seen by the coverage check.
+        covered_cmds = set({'generate'})
```
Comment (pinheadmz) [inline test/functional/test_runner.py:721]: I don't quite understand this. `generate` is not returned by `help` and so I don't think its even added to the coverage list:

https://github.com/bitcoin/bitcoin/blob/bd00d3b1f2036893419d1e8c514a8af2c4e4b1fb/test/functional/test_framework/coverage.py#L96

Both with and without this change, my coverage results were the same:

```
Uncovered RPC commands:
  - pruneblockchain

```

**Comment by pinheadmz** [comment]
Concept ACK

Built and tested locally, all behavior is as expected trying different combinations:

```

**Comment by pinheadmz** [comment]
647  src/bitcoin-cli -regtest -generate
  648  src/bitcoin-cli -regtest generate
  649  src/bitcoin-cli -regtest help generate
  650  src/bitcoin-cli -regtest help -generate
  651  src/bitcoin-cli -regtest -help -generate

**Comment by laanwj** [comment]
> Only thing that caught me as far as interface is the fact that there is help and -help. The extra help menu is a bit of an easter egg? But that's out of scope for this PR I think.

They are principally different on purpose. Like for any UNIX-ish application, `-help` or `--help` shows the help message for the local binary. This includes the different options how to connect to the server.

On the other hand, `help` is a command for the remote server.

File: test/functional/rpc_generate.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+)
+
+
+class RPCGenerateTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+
+    def run_test(self):
+        message = (
+            "generate ( nblocks maxtries ) has been replaced by the -generate "
```
Comment (laanwj) [inline test/functional/rpc_generate.py:20]: Isn't a whole test for what is essentially a stub maybe a bit overkill? :smile:

Comment (jonatack) [reply test/functional/rpc_generate.py:20]: I usually write tests before code; it saved my deliverables so many times and habits die hard :D

Comment (jonatack) [reply test/functional/rpc_generate.py:20]: (OTOH, this TDD definitely a handicap for opening hard-to-test new code in Bitcoin Core and I guess I should unlearn it)

File: test/functional/test_runner.py (hunk 2 above)
Comment (jonatack) [reply test/functional/test_runner.py:721]: @pinheadmz IIRC this was needed to appease a failing check in the Travis CI.

Comment (jonatack) [reply test/functional/test_runner.py:721]: @pinheadmz Yes, it's a hidden command (not in the help, see the added test) as suggested by @laanwj. IIRC this was needed to appease a failing check in the Travis CI that despite the added test it thought there was no coverage.

**Review by pinheadmz — APPROVED** [review]
ACK f0aa8aeea5a183ea44a877255d12db7732f2e0a8

Thanks for your answers to my questions @jonatack and @laanwj 

<details><summary>Show Signature</summary>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

ACK f0aa8aeea5a183ea44a877255d12db7732f2e0a8
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCAAdFiEE5hdzzW4BBA4vG9eM5+KYS2KJyToFAl80EaUACgkQ5+KYS2KJ
yTrkBw//dEwEnhDeko2ScTnzl/4mZSJLR0KG4D//BhWq+bhsSKKslWUQDMdJDYE/
HljPiuButxN01vJOVowqYV6ZGIPEVLiILu/R7OYkU81KLUPtZwsaouT9b0/8opmZ
/uXysTqvUa9UUfrk01vxkU63+viAQdpe5WBaf6dURdBjodRdoxL41r60GhMlacoe
wS75MuzR+pojiyr/Wf9s+t2BKDVO79rhoClhdOo9xiud4kqomdpmyQh2/86T5moe
LgTOjufZGAl/0haI1EAlWmtw0+AjMl6hRnrVUcgblElOJ201HZoE2Qj0L8bkykU1
I890Cu28PFWFeeK6xX3ev/ja6v2iLzrq1VsKmH1WRkEYaOkL8GvH4MJHwuQSlsFS
/h+dxMFjPvmigZMQTuaHDtDrsRGLqMTd6WEYoIC/NKRK+7UjgpGjzSiiRBS4g7iB
pVTTpei9/zjvXwlvr427+ji1kbWt2YKQdqrDeMIHHc5gcsFY9E1vgSKDw94TjVyt
YwOHlYG+ezhULsxNbW/D40ox1veFDkK/I0DkHN0AcUP7RuKTTdAiZ4B6+Fturk0f
3eFNIxf3qilCy5LPfRBcqtQ8GkpVgvcDpsO83cEeffyTIxrtRrPDULRc0upI81Jd
mi6DrhAultLW1xFvAjB8EBTTEVrdg9doVD5vMqHM+CFGm+wnyvM=
=wdk/
-----END PGP SIGNATURE-----
```

pinheadmz's public key is [on keybase](https://keybase.io/pinheadmz/pgp_keys.asc?fingerprint=e61773cd6e01040e2f1bd78ce7e2984b6289c93a)
</details>

File: test/functional/rpc_generate.py (hunk 3 above)
Comment (jonatack) [reply test/functional/rpc_generate.py:20]: Oh, do you mean it should be put in another test instead?

**Comment by jonatack** [comment]
Thanks @pinheadmz!

**Comment by jonatack** [comment]
This has 3 ACKS (2 tested ACKs from @adamjonas and @pinheadmz and a code review ACK from @kallewoof) and is hopefully RFM.
